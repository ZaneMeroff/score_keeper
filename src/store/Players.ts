// Vue & Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Player, PlayerData } from '@/types/players'

@Module
export default class Players extends VuexModule {

    playerData: PlayerData = {}

    get getPlayerData() {
      return this.playerData
    }

    // ------------------------------------------
    //              create player
    // ------------------------------------------

    @Mutation
    createPlayer(payload: Player) {
      Vue.set(this.playerData, payload.id, payload)
    }

    @Action
    action_createPlayer(payload: Player) {
      this.createPlayer(payload)
    }

    // ------------------------------------------
    //         modify target player score
    // ------------------------------------------

    @Mutation
    setPlayerScore(payload: { id: string, score: number }) {
      this.playerData[payload.id].score = payload.score
    }

    @Action
    action_setPlayerScore(payload: { id: string, score: number }) {
      this.setPlayerScore(payload)
    }

    // ------------------------------------------
    //         modify target player name
    // ------------------------------------------

    @Mutation
    setPlayerName(payload: { id: string, name: string }) {
      this.playerData[payload.id].name = payload.name
    }

    @Action
    action_setPlayerName(payload: { id: string, name: string }) {
      this.setPlayerName(payload)
    }

    // ------------------------------------------
    //        zero scores for all players
    // ------------------------------------------

    @Mutation
    zeroScores() {
      for (const player in this.playerData) {
        this.playerData[player].score = 0
      }
    }

    @Action
    action_zeroScores() {
      this.zeroScores()
    }
}