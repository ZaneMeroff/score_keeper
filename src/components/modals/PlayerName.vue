<template>

 <q-dialog v-model="showModal" persistent>
    <q-card class="q-card" :dark="stateSettings.getDarkMode">

      <q-card-section>
        <div class="q-mb-md">

          <!-- header -->
          <div class="text q-mb-sm" :style="stateSettings.getDarkModeText">Edit Name</div>
          <q-separator color="blue" class="q-mb-lg" inset></q-separator>
          
          <!-- text input -->
          <q-input 
            v-model="name" 
            maxlength="20"
            style="font-size: 16px"
            :dark="stateSettings.getDarkMode"
            :rules="[ val => !!val || '* Required' ]"
            @input="onInputChange"
          />
        </div>

        <q-btn-group flat class="q-btn-group">
          <!-- save button -->
          <q-btn
            flat
            class="q-btn"
            color="green"
            id="save-btn"
            size="md"
            :disable="disabled"
            :label="'save'"
            @click="handleSaveBtn"
          />
          <!-- cancel button -->
          <q-btn
            flat
            class="q-btn"
            color="red"
            size="md"
            :label="'cancel'"
            @click="handleCancelBtn"
          />
        </q-btn-group>
      </q-card-section>
    </q-card>
  </q-dialog>

</template>

<script lang="ts">

// Decorators
import { Component, Prop, Vue } from 'vue-property-decorator';

// Vuex
import { statePlayers, stateSettings } from '@/store/index'
import Players from '@/store/Players'
import Settings from '@/store/Settings'

@Component
export default class PlayerName extends Vue {
  
  @Prop(Boolean) private showModal!: boolean
  @Prop(String) private playerId!: string

  statePlayers: Players = statePlayers
  stateSettings: Settings = stateSettings

  name: string = ''
  disabled: boolean = false

  handleSaveBtn() {
    this.statePlayers.action_setPlayerName({ id: this.playerId, name: this.name })
    this.setDefault()
    this.$emit('close')
  }

  handleCancelBtn() {
    this.setDefault()
    this.$emit('close')
  }

  onInputChange() {
    if (this.name.length) this.disabled = false
    else this.disabled = true
  }

  setDefault() {
    if (this.statePlayers.getPlayerData[this.playerId]) {
      this.name = this.statePlayers.getPlayerData[this.playerId].name
    }
  }

  mounted() {
    this.setDefault()
  }
}
</script>

<style scoped>

  .q-card {
    width: 300px;
  }

  .q-btn-group {
    display: inline !important;
  }

  .q-btn {
    width: 50%;
  }

  .text {
    font-size: 16px;
    text-align: center;
  }

</style>