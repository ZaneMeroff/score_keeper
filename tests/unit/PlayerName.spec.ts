import { createLocalVue, shallowMount } from '@vue/test-utils'
import { mountFactory } from '@quasar/quasar-app-extension-testing-unit-jest'
import Players from '@/store/Players'
import PlayerName from '@/components/modals/PlayerName.vue'
import Quasar, * as All from 'quasar'
import uuid from 'uuid/v4'
import Vuex from 'vuex'

// --------- Vuex & Quasar setup -------------

const localVue = createLocalVue()
localVue.use(Vuex, Quasar, {components: All, directives: All, plugins: All})

// ---------------- mocks --------------------

jest.mock('uuid/v4')
uuid.mockImplementation(() => '12345')

// ------------ create player ----------------

// PlayersModule.action_createPlayers(1)

// -------------------------------------------

describe('PlayerName', () => {

  let PlayersModule: Players
  let store: any

  beforeEach(() => {
    store = new Vuex.Store({})
    PlayersModule = new Players({ store, name: 'Players' })
    // PlayersModule.action_createPlayers(1)

    // find out why players exist here but not before tests
    // try connecting PlayerModule to localVue
  })

  describe('snapshots', () => {

    it('should render with default html', () => {
      const propsData = { playerId: '12345', showModal: true }
      const component = shallowMount(PlayerName, { localVue, propsData })
      
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('props', () => {

    describe('showModal', () => {

      it('should accept a boolean', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        expect(component.vm.$props.showModal).toEqual(propsData.showModal)
      })
    })

    describe('playerId', () => {

      it('should accept a string', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        expect(component.vm.$props.playerId).toEqual(propsData.playerId)
      })
    })
  })

  describe('data', () => {

    describe('name', () => {

      it('should have a default value of empty string', () => {

        // console.log(PlayersModule.getPlayerData['12345'])
        // PlayersModule.action_createPlayers(1)

        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        console.log('name: ', component.vm.$data.name)

        const expected = 'Player 1'
        expect(component.vm.$data.name).toEqual(expected)
      })

      it.skip('should accept a string', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const newValue = 'New Name'

        component.setData({ name: newValue })

        expect(component.vm.$data.name).toEqual(newValue)
      })
    })

    describe('disabled', () => {

      it.skip('should have a default value of false', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        expect(component.vm.$data.disabled).toEqual(false)
      })

      it.skip('should accept a boolean', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const newValue = true

        component.setData({ disabled: newValue })

        expect(component.vm.$data.disabled).toEqual(newValue)
      })
    })
  })

  describe('methods', () => {

    describe('handleSaveBtn', () => {
      
      it.skip('should set player name', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const expected = 'Khal Drogo'

        expect(PlayersModule.getPlayerData['12345'].name).toEqual('Player 1')
        
        component.setData({ name: expected });
        (component.vm as any).handleSaveBtn()

        expect(PlayersModule.getPlayerData['12345'].name).toEqual(expected)

        // teardown
        PlayersModule.action_setPlayerName({ id: '12345', name: 'Player 1' })
      })

      it.skip('should call setDefault', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const spy = jest.spyOn((component.vm as any), 'setDefault');

        (component.vm as any).handleSaveBtn()

        expect(spy).toHaveBeenCalledTimes(1)
      })

      it.skip('should emit close event', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        // @ts-ignore
        component.vm.handleSaveBtn()

        const expected = { 'close': [[]] }
        expect(component.emitted()).toEqual(expected)
      })
    })

    describe('handleCancelBtn', () => {
      
      it.skip('should call setDefault', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        // @ts-ignore
        const spy = jest.spyOn(component.vm, 'setDefault')

        // @ts-ignore
        component.vm.handleCancelBtn()

        expect(spy).toHaveBeenCalledTimes(1)
      })

      it.skip('should emit close event', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        // @ts-ignore
        component.vm.handleCancelBtn()

        const expected = { 'close': [[]] }
        expect(component.emitted()).toEqual(expected)
      })
    })

    describe('onInputChange', () => {
      
      it.skip('should set disabled to true if name.length=false', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })

        component.setData({ name: '' })
        // @ts-ignore
        component.vm.onInputChange()

        expect(component.vm.$data.disabled).toEqual(true)
      })

      it.skip('should set disabled to false if name.length=true', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const newValue = 'Sample Name'

        component.setData({ name: newValue })
        // @ts-ignore
        component.vm.onInputChange()

        expect(component.vm.$data.disabled).toEqual(false)
      })
    })

    describe('setDefault', () => {
      
      it.skip('should set name correctly if player exists', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const playerName = 'Player 1'

        component.setData({ name: playerName })
        // @ts-ignore
        component.vm.setDefault()

        expect(component.vm.$data.name).toEqual(playerName)
      })
    })
  })

  describe('lifecycle methods', () => {

    describe('mounted', () => {

      it.skip('should call setDefault on mount', () => {
        const propsData = { playerId: '12345', showModal: true }
        // @ts-ignore
        const spy = jest.spyOn(PlayerName.options.methods, 'setDefault')
        const component = shallowMount(PlayerName, { localVue, propsData })

        expect(spy).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('user interactions', () => {

    describe('events', () => {

      it.skip('should call handeSaveBtn when save btn is clicked', async () => {
        // @ts-ignore
        const spy = jest.spyOn(PlayerName.options.methods as any, 'handleSaveBtn')

        // @ts-ignore
        const factory = mountFactory(PlayerName, { quasar: All, mount: { type: 'full' } })

        const wrapper = factory({ playerId: '12345', showModal: true })

        await wrapper.find('#save-btn').trigger('click')
        // ************************************************
        // ************************************************
        // ***************  incomplete  *******************
        // ************************************************
        // ************************************************
        expect(spy).toHaveBeenCalledTimes(1)
      })
    })
  })
})
