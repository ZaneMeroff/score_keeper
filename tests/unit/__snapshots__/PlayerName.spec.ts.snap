// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerName snapshots should render with default html 1`] = `
<q-dialog
  persistent=""
>
  <q-card
    class="q-card"
    dark="true"
  >
    <q-card-section>
      <div
        class="q-mb-md"
      >
        <div
          class="text q-mb-sm"
          style="color: rgb(255, 255, 255);"
        >
          Edit Name
        </div>
         
        <q-separator
          class="q-mb-lg"
          color="blue"
          inset=""
        />
         
        <q-input
          dark="true"
          maxlength="20"
          rules="function (val) { return !!val || '* Required'; }"
          style="font-size: 16px;"
        />
      </div>
       
      <q-btn-group
        class="q-btn-group"
        flat=""
      >
        <q-btn
          class="q-btn"
          color="green"
          flat=""
          id="save-btn"
          label="save"
          size="md"
        />
         
        <q-btn
          class="q-btn"
          color="red"
          flat=""
          label="cancel"
          size="md"
        />
      </q-btn-group>
    </q-card-section>
  </q-card>
</q-dialog>
`;
